synapse/handlers/message.py

```diff
@@ -949,14 +950,24 @@ async def create_new_client_event(
         else:
             prev_event_ids = await self.store.get_prev_events_for_room(builder.room_id)
 
-        # we now ought to have some prev_events (unless it's a create event).
-        #
-        # do a quick sanity check here, rather than waiting until we've created the
+        # Do a quick sanity check here, rather than waiting until we've created the
         # event and then try to auth it (which fails with a somewhat confusing "No
         # create event in auth events")
-        assert (
-            builder.type == EventTypes.Create or len(prev_event_ids) > 0
-        ), "Attempting to create an event with no prev_events"
+        room_version_obj = await self.store.get_room_version(builder.room_id)
+        if room_version_obj.msc2716_empty_prev_events:
```

Comment:
Does this actually require a new room version? (discussed at https://github.com/matrix-org/synapse/pull/11114#discussion_r733475645)

My thinking is that this code is only for creating events, not accepting events. So technically any other homeserver nowadays can create events with no `prev_events` now and it would work.

Reply:
I am not that surprised that we accept outliers with empty prev-events, but we'll need special handling for the insertion event to be accepted, i.e. we'll need some sort of check like "this is an insertion event so we need to go and fetch the state rather than trying to calculate it". Though that can be part of the history import MSC I guess

Reply:
> we'll need special handling for the insertion event to be accepted, i.e. we'll need some sort of check like "this is an insertion event so we need to go and fetch the state rather than trying to calculate it"

Why is this the case? It seems to work in my Complement tests without any of this 🤔

Reply:
Spooky! :ghost:  Is it doing something silly like dropping the event with no extremities and then doing a `/state` on the event that references it?

Reply:
I'm not sure yet. If you really want me to dive into it, I can ⛳